### PR TITLE
Upgrade to Elixir 1.8 and Ecto 3.0

### LINKED ISSUE #489: Upgrade to Elixir 1.8

### PR DESCRIPTION
**Pull request by unnawut**
Issue/Task Number: #489
Closes #489

# Overview

This PR upgrades the entire eWallet server to Elixir 1.8 and Ecto 3.0 plus a few more smaller dependencies.

⚠️ **Change `.circleci/config.yml: executors.builder.docker.image` back to `stable` (or `latest`?) when the builder image has been upgraded to Elixir 1.8, i.e. https://github.com/omisego-images/docker-ewallet/pull/1 is merged.**

# Changes

- Upgraded to Elixir 1.8 and Ecto 3.0
- Upgraded to the latest versions of `:quantum`, `:credo`, `:dialyxir`, `:ex_doc`, `:junit_formatter`, `:sentry`
- Bumped all sub-apps' minimum Elixir version to 1.8
- Fix various breaking changes resulted from Ecto 2.2 & 3.0 changes.
  - `Multi.run/3` now expects the passed callback to receive a repo
  - `Query.join/5` functions no longer wrap fragments with parentheses
  - `Ecto.DateTime` removed
  - `:ecto_autogenerate` expecting a list of field name instead of a single field name
  - Contraint names are now passed into validation error metadata
  - `Changeset.validate_number/3`'s errors now returning `:kind`, e.g. `[validation: number, kind: :greater_than, number: 0]`
  - Plural time units are removed, e.g. `:seconds` becomes `:second`
  - `Repo.update_all/3` no longer supports `returning: true`
  - `Repo.update/3` no longer return unrelated attributes
  - `Changeset.validate_*` now removes invalid fields from `changeset.changes`
  - Timestamps are operated without milli/microseconds by default (why Ecto whyyyy 😱)
  - `Repo.preload/3` no longer re-preloads if the association has already been loaded before
  - `Query.preload/3` in a stream query is no longer supported
- Fix `ActivityLogger` saving records with no changes

# Implementation Details

All the changes are pretty straightforward and only cover changes that resulted from the new versions.

One noteworthy change is the shifted default from microseconds to seconds. You'll see in this PR that **existing migrations are edited from `:naive_datetime` to `:naive_datetime_usec` instead of creating a new migration.**

I did this with assumption of these possible scenarios:

1. **Instances setup after this PR:** All migration files are now `:naive_datetime_usec`, which mean the schemas will be created exactly as in previous versions with datetime in microseconds.

2. **Instances setup before this PR:** The already-migrated datetime fields are already in microseconds. The `:naive_datetime_usec` ensures that rolling back and forward even across those migrations will still respect microseconds that used to be.

Hence I think it makes the most sense to edit existing migrations to reflect the actual semantics of the database rather than keeping the literal migration code.

# Usage

## Dev environment

Anyone who has taken code that contains this PR must do the following:

1. Upgrade your Elixir version to `v1.8.x`. Follow the instructions at [Installing Elixir](https://elixir-lang.org/install.html).
2. Make sure that the correct Elixir version is being used by running:
      ```
      $ elixir --version
      Elixir 1.8.1 (compiled with OTP xx)
      ```
3. Clean the dev, test and dependency builds by running:
      ```
      mix clean && mix deps.clean -all && MIX_ENV=test mix clean
      ```
4. Retrieve the new dependencies as specified in mix.lock:
      ```
      mix deps.get
      ```
5. Just to be safe, compile the source code manually once:
      ```
      mix compile
      ```

Upgrade is complete! Continue saving the world!

## Production environment

We'll need to wait for https://github.com/omisego-images/docker-ewallet to support Elixir 1.8 before proceeding with this PR.

# Impact

**For build to pass, https://github.com/omisego-images/docker-ewallet and the whole CI workflow must be running or somehow supporting Elixir 1.8.**

Other than that, no changes to the DB schema or API specs.

### CODE REVIEW

**Comment by unnawut** [comment]
@sirn This PR is now passing on my local machine. Now I'll need your help configuring Elixir 1.8 on the CI 🙇

**Comment by sirn** [comment]
I think we can have v1.1 branch uses `omisegoimages/ewallet-builder:v1.1`. That would be the easiest way, but we have to be careful when merging to not include that particular v1.1 change commit. Or include that, but make some change on top of it.

**Comment by unnawut** [comment]
Forgot that we still need to fix the build error:

```
==> erlex
Compiling 2 files (.erl)
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:59: illegal pattern
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:60: variable 'Error' is unbound
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:60: variable 'Stacktrace' is unbound
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:63: variable 'Stacktrace' unsafe in 'try' (line 60)
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:64: variable 'Error' unsafe in 'try' (line 60)
/usr/local/otp/lib/erlang/lib/parsetools-2.1.8/include/yeccpre.hrl:64: variable 'Stacktrace' unsafe in 'try' (line 60)
could not compile dependency :erlex, "mix compile" failed. You can recompile this dependency with "mix deps.compile erlex", update it with "mix deps.update erlex" or clean it with "mix deps.clean erlex"
make: *** [Makefile:73: build-test] Error 1
Exited with code 2
```

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ version: 2.1
 executors:
   builder:
     docker:
-      - image: omisegoimages/ewallet-builder:stable
+      - image: omisegoimages/ewallet-builder:dev-80264b0
```
Comment (sirn) [inline .circleci/config.yml]: You can change this back to stable once we merge the 1.8 image :-)

**Comment by sirn** [comment]
Can you also update .tool-versions?

**Comment by unnawut** [comment]
**Merge instructions:**
- [ ] Merge https://github.com/omisego-images/docker-ewallet/pull/1 and wait for the build to succeed. `ewallet-builder:latest` image should be produced
- [ ] Tag `ewallet-builder:latest` with `ewallet-builder:v1.2`
- [ ] Update this PR to use `omisegoimages/ewallet-builder:v1.2` (in `./circleci/config.yml`, `docker-gen.sh`, `Makefile`)
- [ ] Update all `Mixfile`'s project.version to `v1.2.0-dev`
- [ ] ~Update `docker-compose.override.yml` to `omiseo/ewallet:stable`~ (not until `v1.1` is released and tagged `stable`)
- [ ] ~Update `docker-compose.yml` to `omiseo/ewallet:stable`~ (not until `v1.1` is released and tagged `stable`)
- [ ] Merge!

**Pull request merged**